README.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddYMap

A widget for ManagerManager plugin allowing Yandex Maps integration.


## Requires

* PHP >= 5.4
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.ru/modx/managermanager) >= 0.7


## Installation

To install you must unzip the archive to `/assets/plungins/managermanager/widgets/ddymap/`.


You may also read this documentation:
* [(MODX)EvolutionCMS.plugins.ManagerManager](https://code.divandesign.ru/modx/managermanager).
* [(MODX)EvolutionCMS.modules.ddMMEditor](https://code.divandesign.ru/modx/ddmmeditor).


## Parameters description

* `$params`
	* Desctription: The object of parameters.
	* Valid values:
		* `stdClass`
		* `arrayAssociative`
	* **Required**
	
* `$params->fields`
	* Desctription: TV names to which the widget is applied.
	* Valid values:
		* `stringCommaSeparated`
		* `array`
	* **Required**
	
* `$params->fields[$i]`
	* Desctription: The name of the document TV.
	* Valid values: `string`
	* **Required**
	
* `$params->mapWidth`
	* Desctription: Width of the map container.
	* Valid values:
		* `integer`
		* `'auto'`
	* Default value: `'auto'`
	
* `$params->mapHeight`
	* Desctription: Height of the map container.
	* Valid values: `integer`
	* Default value: `400`
	
* `$params->hideOriginalInput`
	* Desctription: Original coordinates field hiding status.
	* Valid values: `boolean`
	* Default value: `true`
	
* `$params->defaultZoom`
	* Desctription: Default map zoom.
	* Valid values: `integer`
	* Default value: `15`
	
* `$params->defaultPosition`
	* Desctription: Default map position when a document field is empty.
	* Valid values: `stringCommaSeparated`
	* Default value: `'55.20432131317031,61.28999948501182'`
	
* `$params->defaultPosition[0]`
	* Desctription: Latitude.
	* Valid values: `float`
	* Default value: `55.20432131317031`
	
* `$params->defaultPosition[1]`
	* Desctription: Longitude.
	* Valid values: `float`
	* Default value: `61.28999948501182`
	
* `$params->roles`
	* Desctription: The CMS user roles that the widget is applied to.
	* Valid values:
		* `array`
		* `stringCommaSeparated`
		* `''` — when this parameter is empty then widget is applied to the all roles
	* Default value: `''`
	
* `$params->roles[$i]`
	* Desctription: CMS user role ID.
	* Valid values: `integer`
	* **Required**
	
* `$params->templates`
	* Desctription: Document templates IDs for which the widget is applied to.
	* Valid values:
		* `array`
		* `stringCommaSeparated`
		* `''` — empty value means the widget is applying to all templates
	* Default value: `''`
	
* `$params->templates[$i]`
	* Desctription: Template ID.
	* Valid values: `integer`
	* **Required**


## CMS events

* `OnDocFormPrerender`
* `OnDocFormRender`


## Examples


### Yandex Maps widget connection to the “LatLng” TV

```php
mm_ddYMap([
	'fields' => 'LatLng'
]);
```


## Links

* [Home page](https://code.divandesign.ru/modx/mm_ddmap)
* [Telegram chat](https://t.me/dd_code)
* [Packagist](https://packagist.org/packages/dd/evolutioncms-plugins-managermanager-mm_ddmap)
* [GitHub](https://github.com/DivanDesign/EvolutionCMS.plugins.ManagerManager.mm_ddMap)


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />

CHANGELOG.md
# (MODX)EvolutionCMS.plugins.ManagerManager.mm_ddMap changelog


## Version 1.6 (2016-11-25)
* \* The widget works correctly with HTTPS. Many thanks to [@byscrimm](https://github.com/byscrimm).
* \* Parameters:
	* \* The widget now uses “named” parameters, see README → Examples (with backward compatibility).
	* \+ `$params->defaultZoom`: The new parameter. It allows to set default map zoom.
	* \+ `$params->defaultPosition`: The new parameter. It allows to set default map position when a document field is empty. Many thanks to [MrSwed](https://github.com/MrSwed) and all caring people.
* \* The map section header now contains all HTML tags from the original field title.
* \* jQuery.ddYMap has been updated to 1.4.
* \* Attention! PHP >= 5.4 is required.
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.7 is required.


## Version 1.5b (2015-05-08)
* \* The 2.1 version of Yandex Maps  API is used. The plugin jQuery.ddYMap 1.3.1 is used for map rendering.
* \* The `div.sectionBody` element has no longer the `tmplvars` class while creating a section.
* \* The `$w` and `$h` parameters have been renamed as `$width` and `$height` respectively.


## Version 1.4.3 (2013-12-10)
* \* Just minor changes.
* \* Unix format is currently used for line delimiters.


## Version 1.4.2 (2013-11-15)
* \* The calls of the `includeJs` function were replaced by `includeJsCss`.
* \* The including of required js and css files is currently being realized while occuring the event `OnDocFormPrerender`. The files are included as html.
* \* The file `jquery.ddMM.mm_ddYMap-1.0.1.js` has been renamed as `jquery.ddMM.mm_ddYMap.js`.
* \* The `$.ddMM.mm_ddYMap` plugin has been updated to 1.0.2:
	* \* The processing of the map search form submitting was removed because of its uselessness since the submitting bug of the main document form had been fixed.


## Version 1.4.1 (2013-10-25)
* \* Minor changes to ensure compatibility with (MODX)EvolutionCMS.plugins.ManagerManager 0.6.


## Version 1.4 (2013-10-04)
* \* The template id of a current document is set equally to `$mm_current_page['template']`.
* \* The js code was revised completely and became a stand-alone file. It is convenient because of php code cleanness and it shortens the amount of code of a document edit frame.
* \* Maps are rendered in the same order as TV names are supplied in `$tvs` parameter.
* \* Attention! (MODX)EvolutionCMS.plugins.ManagerManager >= 0.6 is required.


## Version 1.3.1 (2013-06-08)
* \* The error that occurred in Opera and IE was eliminated.


## Version 1.3 (2013-05-18)
* \+ Search control has been added.


## Version 1.2.2 (2012-11-14)
* \* The bug with no coordinates updating while dropping the placemark has been fixed.


## Version 1.2.1 (2012-11-13)
* \* The bug with no widget initialization while creating a new document has been fixed.
* \* Minor code changes.


## Version 1.2 (2012-05-02)
* \+ The code has been rewritten to provide compatibility with the second version of API Yandex Maps.
* \- The parameter `key` has been deleted (because of its irrelevance).


## Version 1.1 (2012-04-16)
* \+ The parameter `hideField` that allows showing/hiding original coordinates pole has been added.
* \* Minor changes in js-code.


## Version 1.0.1 (2012-01-13)
* \* Modx check for event `OnDocFormRender` has been added.


<link rel="stylesheet" type="text/css" href="https://raw.githack.com/DivanDesign/CSS.ddMarkdown/master/style.min.css" />
<style>ul{list-style:none;}</style>

ddymap.php
<?php
/**
 * mm_ddYMap
 * @version 1.6 (2016-11-25)
 * 
 * @see README.md
 * 
 * @link https://code.divandesign.biz/modx/mm_ddymap/1.6
 * 
 * @copyright 2012–2016 DivanDesign {@link https://DivanDesign.biz }
 */

function mm_ddYMap($params){
	//For backward compatibility
	if (
		!is_array($params) &&
		!is_object($params)
	){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'fields',
				'roles',
				'templates',
				'mapWidth',
				'mapHeight',
				'hideOriginalInput'
			]
		]);
	}
	
	//Defaults
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			(object) [
//	 			'fields' => '',
				'mapWidth' => 'auto',
				'mapHeight' => 400,
				'hideOriginalInput' => true,
				'defaultZoom' => '',
				'defaultPosition' => '',
				'roles' => '',
				'templates' => '',
			],
			$params
		]
	]);
	
	if (
		!useThisRule(
			$params->roles,
			$params->templates
		)
	){
		return;
	}
	
	global $modx;
	
	if ($modx->Event->name == 'OnDocFormPrerender'){
		//The jQuery.ddYMap library including
		$output .= includeJsCss(
			$modx->config['site_url'] . 'assets/plugins/managermanager/widgets/ddymap/jQuery.ddMap-2.0.min.js',
			'html',
			'jQuery.ddMap',
			'2.0'
		);
		//The main js file including
		$output .= includeJsCss(
			$modx->config['site_url'] . 'assets/plugins/managermanager/widgets/ddymap/jQuery.ddMM.mm_ddMap.js',
			'html',
			'jQuery.ddMM.mm_ddMap',
			'2.0'
		);
		
		$modx->Event->output($output);
	}elseif ($modx->Event->name == 'OnDocFormRender'){
		if (is_array($params->fields)){
			$params->fields = implode(
				',',
				$params->fields
			);
		}
		
		$output = 
'
//---------- mm_ddYMap :: Begin -----
$j.ddMM.getFieldElems({fields: "' . $params->fields . '"}).mm_ddMap({
	hideField: ' . intval($params->hideOriginalInput) . ',
	width: "' . $params->mapWidth . '",
	height: "' . $params->mapHeight . '"' . (
		!empty($params->defaultZoom) ?
		', defaultZoom: ' . intval($params->defaultZoom) :
		''
	) . (
		!empty($params->defaultPosition) ?
		', defaultPosition: "' . $params->defaultPosition . '"' :
		''
	) . '
});
//---------- mm_ddYMap :: End -----
';
		
		$modx->Event->output($output);
	}
}
?>

jQuery.ddMM.mm_ddMap.js
/**
 * jQuery.ddMM.mm_ddMap
 * @version 2.0.1 (2023-05-24)
 * 
 * @uses Yandex.Maps 2.1
 * @uses jQuery 1.10.2
 * @uses jQuery.ddMM 1.0
 * @uses jQuery.ddMap 2.0
 * 
 * @copyright 2013–2023 Ronef {@link https://Ronef.ru }
 */

(function($){
$.ddMM.mm_ddMap = {
	//Параметры по умолчанию
	defaults: {
		//Нужно ли скрывать оригинальное поле
		hideField: true,
		//Ширина контейнера с картой
		width: 'auto',
		//Высота контейнера с картой
		height: 400,
		//Default map zoom
		defaultZoom: 15,
		//Default map position when field has no value
		defaultPosition: '55.20432131317031,61.28999948501182'
	},
	
	/**
	 * @method init
	 * @version 3.0.2 (2023-05-23)
	 * 
	 * @desc Инициализация карты.
	 * 
	 * @param elem {objectPlain} — The parameters.
	 * @param elem.position {array} — Position.
	 * @param elem.position[0] {float} — Lat.
	 * @param elem.position[1] {float} — Lng.
	 * @param elem.$map {jQuery} — Map container.
	 * @param elem.$coordInput {jQuery} — Coordinates field.
	 * @param elem.defaultZoom {integer} — Default map zoom.
	 * 
	 * @returns {void}
	 */
	init: function(elem){
		//После инициализации карты
		elem.$map
			.on(
				'ddAfterInit',
				function(){
					//Объект карты
					var
						map = elem.$map.data('ddMap').map,
						//Контрол поиска
						serachControl = new ymaps.control.SearchControl({
							options: {
								useMapBounds: true,
								noPlacemark: true,
								maxWidth: 400
							}
						}),
						//Метка.
						//TODO: Это очень странно, но похоже, что «map.geoObjects.get(0)» возвращает «GeoObjectCollection» вместо «Placemark», потому приходится ещё раз делать «get(0)».
						placemark = map.geoObjects.get(0).get(0)
					;
					
					//При выборе результата поиска
					serachControl.events.add(
						'resultselect',
						function(event){
							var coords = event.originalEvent.target.getResultsArray()[0].geometry.getCoordinates();
							
							//Переместим куда надо маркер
							placemark.geometry.setCoordinates(coords);
							
							//Запишем значение в оригинальное поле
							elem.$coordInput.val(coords[0] + ',' + coords[1]);
						}
					);
					
					map.controls.add(serachControl);
					
					//При клике по карте меняем координаты метки
					map.events.add(
						'click',
						function(event){
							var coords = event.get('coords');
							
							placemark.geometry.setCoordinates([coords[0], coords[1]]);
							
							elem.$coordInput.val(coords[0] + ',' + coords[1]);
						}
					);
					
					//Перетаскивание метки
					placemark.events.add(
						'dragend',
						function(event){
							var coords = placemark.geometry.getCoordinates();
							
							elem.$coordInput.val(coords[0] + ',' + coords[1]);
						}
					);
				}
			)
			.ddMap({
				markers: elem.position,
				markerOptions: {draggable: true},
				defaultZoom: elem.defaultZoom,
			})
		;
	}
};

/**
 * jQuery.fn.mm_ddMap
 * @version 2.0 (2023-05-24)
 * 
 * @desc Делает карту.
 * 
 * @uses $.ddMM.mm_ddMap
 * 
 * @param [params] {objectPlain} — Параметры передаются в виде plain object.
 * @param [params.hideField=true] {boolean} — Нужно ли скрывать оригинальное поле.
 * @param [params.width='auto'] {integer|'auto'} — Ширина контейнера с картой.
 * @param [params.height=400] {integer} — Высота контейнера с картой.
 * @param [params.defaultZoom] {integer} — Default map zoom.
 * 
 * @copyright 2013–2023 Ronef {@link https://Ronef.ru }
 */
$.fn.mm_ddMap = function(params){
	//Обрабатываем параметры
	params = $.extend(
		{},
		$.ddMM.mm_ddMap.defaults,
		params || {}
	);
	
	//Если ширина является числом
	if ($.isNumeric(params.width)){
		//Допишем пиксели
		params.width += 'px';
	}
	
	return $(this).each(function(){
		var elem = {};
		
		//TV с координатами
		elem.$coordInput = $(this);
		//Координаты
		elem.position = elem.$coordInput.val();
		//Default map zoom
		elem.defaultZoom = params.defaultZoom;
		
		//Родитель
		var
			$coordInputParent = elem.$coordInput.parents('tr:first'),
			//Запоминаем название поля
			sectionName = $coordInputParent.find('.warning').html(),
			//Контейнер для карты
			$sectionContainer = $('<div class="sectionHeader">' + sectionName + '</div><div class="sectionBody"></div>')
		;
		
		elem.$map = $('<div style="width: ' + params.width + '; height: ' + params.height + 'px; position: relative; border: 1px solid #c3c3c3;"></div>');
		elem.$map.appendTo($sectionContainer.filter('.sectionBody'));
		
		//Добавляем контейнер
		elem.$coordInput.parents('.tab-page:first').append($sectionContainer);
		
		//Скрываем родителя и разделитель
		$coordInputParent.hide().prev('tr').hide();
		
		//Если скрывать не надо, засовываем перед картой
		if (!params.hideField){
		 	elem.$coordInput.insertBefore(elem.$map);
		}
		
		//Если координаты не заданны, то задаём дефолт
		if ($.trim(elem.position) == ''){
			elem.position = params.defaultPosition;
		}
		
		//Разбиваем координаты
		elem.position = elem.position.split(',');
		
		//Инициализируем
		$.ddMM.mm_ddMap.init(elem);
	});
};
})(jQuery);

jQuery.ddMM.mm_ddYMap.js
/**
 * jQuery.ddMM.mm_ddYMap
 * @version 1.1.7 (2023-05-23)
 * 
 * @uses Yandex.Maps 2.1
 * @uses jQuery 1.10.2
 * @uses jQuery.ddMM 1.0
 * @uses jQuery.ddMap 2.0
 * 
 * @copyright 2013–2023 [DivanDesign]{@link https://DivanDesign.biz }
 */

(function($){
$.ddMM.mm_ddYMap = {
	//Параметры по умолчанию
	defaults: {
		//Нужно ли скрывать оригинальное поле
		hideField: true,
		//Ширина контейнера с картой
		width: 'auto',
		//Высота контейнера с картой
		height: 400,
		//Default map zoom
		defaultZoom: 15,
		//Default map position when field has no value
		defaultPosition: '55.20432131317031,61.28999948501182'
	},
	
	/**
	 * @method init
	 * @version 3.0.2 (2023-05-23)
	 * 
	 * @desc Инициализация карты.
	 * 
	 * @param elem {objectPlain} — The parameters.
	 * @param elem.position {array} — Position.
	 * @param elem.position[0] {float} — Lat.
	 * @param elem.position[1] {float} — Lng.
	 * @param elem.$map {jQuery} — Map container.
	 * @param elem.$coordInput {jQuery} — Coordinates field.
	 * @param elem.defaultZoom {integer} — Default map zoom.
	 * 
	 * @returns {void}
	 */
	init: function(elem){
		//После инициализации карты
		elem.$map
			.on(
				'ddAfterInit',
				function(){
					//Объект карты
					var
						map = elem.$map.data('ddMap').map,
						//Контрол поиска
						serachControl = new ymaps.control.SearchControl({
							options: {
								useMapBounds: true,
								noPlacemark: true,
								maxWidth: 400
							}
						}),
						//Метка.
						//TODO: Это очень странно, но похоже, что «map.geoObjects.get(0)» возвращает «GeoObjectCollection» вместо «Placemark», потому приходится ещё раз делать «get(0)».
						placemark = map.geoObjects.get(0).get(0)
					;
					
					//При выборе результата поиска
					serachControl.events.add(
						'resultselect',
						function(event){
							var coords = event.originalEvent.target.getResultsArray()[0].geometry.getCoordinates();
							
							//Переместим куда надо маркер
							placemark.geometry.setCoordinates(coords);
							
							//Запишем значение в оригинальное поле
							elem.$coordInput.val(coords[0] + ',' + coords[1]);
						}
					);
					
					map.controls.add(serachControl);
					
					//При клике по карте меняем координаты метки
					map.events.add(
						'click',
						function(event){
							var coords = event.get('coords');
							
							placemark.geometry.setCoordinates([coords[0], coords[1]]);
							
							elem.$coordInput.val(coords[0] + ',' + coords[1]);
						}
					);
					
					//Перетаскивание метки
					placemark.events.add(
						'dragend',
						function(event){
							var coords = placemark.geometry.getCoordinates();
							
							elem.$coordInput.val(coords[0] + ',' + coords[1]);
						}
					);
				}
			)
			.ddMap({
				markers: elem.position,
				markerOptions: {draggable: true},
				defaultZoom: elem.defaultZoom,
			})
		;
	}
};

/**
 * jQuery.fn.mm_ddYMap
 * @version 1.1.5 (2023-05-23)
 * 
 * @desc Делает карту.
 * 
 * @uses $.ddMM.mm_ddYMap
 * 
 * @param [params] {objectPlain} — Параметры передаются в виде plain object.
 * @param [params.hideField=true] {boolean} — Нужно ли скрывать оригинальное поле.
 * @param [params.width='auto'] {integer|'auto'} — Ширина контейнера с картой.
 * @param [params.height=400] {integer} — Высота контейнера с картой.
 * @param [params.defaultZoom] {integer} — Default map zoom.
 * 
 * @copyright 2013–2016 [DivanDesign]{@link http://www.DivanDesign.biz }
 */
$.fn.mm_ddYMap = function(params){
	//Обрабатываем параметры
	params = $.extend(
		{},
		$.ddMM.mm_ddYMap.defaults,
		params || {}
	);
	
	//Если ширина является числом
	if ($.isNumeric(params.width)){
		//Допишем пиксели
		params.width += 'px';
	}
	
	return $(this).each(function(){
		var elem = {};
		
		//TV с координатами
		elem.$coordInput = $(this);
		//Координаты
		elem.position = elem.$coordInput.val();
		//Default map zoom
		elem.defaultZoom = params.defaultZoom;
		
		//Родитель
		var
			$coordInputParent = elem.$coordInput.parents('tr:first'),
			//Запоминаем название поля
			sectionName = $coordInputParent.find('.warning').html(),
			//Контейнер для карты
			$sectionContainer = $('<div class="sectionHeader">' + sectionName + '</div><div class="sectionBody"></div>')
		;
		
		elem.$map = $('<div style="width: ' + params.width + '; height: ' + params.height + 'px; position: relative; border: 1px solid #c3c3c3;"></div>');
		elem.$map.appendTo($sectionContainer.filter('.sectionBody'));
		
		//Добавляем контейнер
		elem.$coordInput.parents('.tab-page:first').append($sectionContainer);
		
		//Скрываем родителя и разделитель
		$coordInputParent.hide().prev('tr').hide();
		
		//Если скрывать не надо, засовываем перед картой
		if (!params.hideField){
		 	elem.$coordInput.insertBefore(elem.$map);
		}
		
		//Если координаты не заданны, то задаём дефолт
		if ($.trim(elem.position) == ''){
			elem.position = params.defaultPosition;
		}
		
		//Разбиваем координаты
		elem.position = elem.position.split(',');
		
		//Инициализируем
		$.ddMM.mm_ddYMap.init(elem);
	});
};
})(jQuery);

ddymap/jquery.ddMM.mm_ddYMap.js
/**
 * jQuery ddMM.mm_ddYMap Plugin
 * @version 1.1 (2015-05-07)
 * 
 * @uses Yandex.Maps 2.1.
 * @uses jQuery 1.10.2.
 * @uses jQuery.ddMM 1.0.
 * @uses jQuery.ddYMap 1.3.1.
 * 
 * @copyright 2015, DivanDesign
 * http://www.DivanDesign.biz
 */

(function($){
$.ddMM.mm_ddYMap = {
	//Параметры по умолчанию
	defaults: {
		//Нужно ли скрывать оригинальное поле
		hideField: true,
		//Ширина контейнера с картой
		width: 'auto',
		//Высота контейнера с картой
		height: 400,
		// Координаты по умолчанию. Москва.
		LngLat: '55.751574295017306,37.61860046759789'
	},
	//Инициализация карты
	init: function(elem){
		var $mapElement = $('#ddYMap' + elem.id);
		
		//После инициализации карты
		$mapElement.on('ddAfterInit', function(){
			//Объект карты
			var map = $mapElement.data('ddYMap').map,
				//Контрол поиска
				serachControl = new ymaps.control.SearchControl({
					options: {
						useMapBounds: true,
						noPlacemark: true,
						maxWidth: 400
					}
				}),
				//Метка.
				//TODO: Это очень странно, но похоже, что «map.geoObjects.get(0)» возвращает «GeoObjectCollection» вместо «Placemark», потому приходится ещё раз делать «get(0)».
				placemark = map.geoObjects.get(0).get(0);
			
			//При выборе результата поиска
			serachControl.events.add('resultselect', function(event){
				var coords = event.originalEvent.target.getResultsArray()[0].geometry.getCoordinates();
				
				//Переместим куда надо маркер
				placemark.geometry.setCoordinates(coords);
				
				//Запишем значение в оригинальное поле
				elem.$elem.val(coords[0] + ',' + coords[1]);
			});
			
			map.controls.add(serachControl);
			
			//При клике по карте меняем координаты метки
			map.events.add('click', function(event){
				var coords = event.get('coords');
				
				placemark.geometry.setCoordinates([coords[0], coords[1]]);
				
				elem.$elem.val(coords[0] + ',' + coords[1]);
			});
			
			//Перетаскивание метки
			placemark.events.add('dragend', function(event){
				var coords = placemark.geometry.getCoordinates();
				
				elem.$elem.val(coords[0] + ',' + coords[1]);
			});
		}).ddYMap({
			placemarks: elem.LngLat,
			placemarkOptions: {draggable: true}
		});
	}
};

/**
 * jQuery.fn.mm_ddYMap Plugin
 * @version 1.0.2 (2015-05-07)
 * 
 * @desc Делает карту.
 * 
 * @uses $.ddMM.mm_ddYMap
 * 
 * @param params {plain object} - Параметры передаются в виде plain object.
 * @param params.hideField {boolean} - Нужно ли скрывать оригинальное поле. Default: true.
 * @param params.width {integer; 'auto'} - Ширина контейнера с картой. Default: 'auto'.
 * @param params.height {integer} - Высота контейнера с картой. Default: 400.
 * 
 * @copyright 2015, DivanDesign
 * http://www.DivanDesign.biz
 */
$.fn.mm_ddYMap = function(params){
	//Обрабатываем параметры
	params = $.extend({}, $.ddMM.mm_ddYMap.defaults, params || {});
	
	//Если ширина является числом
	if ($.isNumeric(params.width)){
		//Допишем пиксели
		params.width += 'px';
	}
	
	return $(this).each(function(){
		var elem = {};
		
		//TV с координатами
		elem.$elem = $(this);
		//ID оригинальной TV
		elem.id = elem.$elem.attr('id');
		//Координаты
		elem.LngLat = elem.$elem.val();
		
		//Родитель
		var	$elemParent = elem.$elem.parents('tr:first'),
			//Запоминаем название поля
			sectionName = $elemParent.find('.warning').text(),
			//Контейнер для карты
			$sectionContainer = $('<div class="sectionHeader">' + sectionName + '</div><div class="sectionBody"><div id="ddYMap' + elem.id + '" style="width: ' + params.width + '; height: ' + params.height + 'px; position: relative; border: 1px solid #c3c3c3;"></div></div>'),
			$YMap = $sectionContainer.find('#ddYMap' + elem.id);
		
		//Добавляем контейнер
		elem.$elem.parents('.tab-page:first').append($sectionContainer);
		
		//Скрываем родителя и разделитель
		$elemParent.hide().prev('tr').hide();
		
		//Если скрывать не надо, засовываем перед картой
		if (!params.hideField){
		 	elem.$elem.insertBefore($YMap);
		}
		
		//Если координаты не заданны, то задаём дефолт
		if ($.trim(elem.LngLat) == ''){
			elem.LngLat = params.LngLat;
		}
		
		//Разбиваем координаты
		elem.LngLat = elem.LngLat.split(',');
		
		//Инициализируем
		$.ddMM.mm_ddYMap.init(elem);
	});
};
})(jQuery);

mm_ddmap.php
<?php
/**
 * mm_ddMap
 * @version 1.6 (2016-11-25)
 * 
 * @see README.md
 * 
 * @link https://code.divandesign.ru/modx/mm_ddmap
 * 
 * @copyright 2012–2016 Ronef {@link https://Ronef.ru }
 */

function mm_ddMap($params){
	//For backward compatibility
	if (
		!is_array($params) &&
		!is_object($params)
	){
		//Convert ordered list of params to named
		$params = \ddTools::orderedParamsToNamed([
			'paramsList' => func_get_args(),
			'compliance' => [
				'fields',
				'roles',
				'templates',
				'mapWidth',
				'mapHeight',
				'hideOriginalInput'
			]
		]);
	}
	
	//Defaults
	$params = \DDTools\ObjectTools::extend([
		'objects' => [
			(object) [
//	 			'fields' => '',
				'mapWidth' => 'auto',
				'mapHeight' => 400,
				'hideOriginalInput' => true,
				'defaultZoom' => '',
				'defaultPosition' => '',
				'roles' => '',
				'templates' => '',
			],
			$params
		]
	]);
	
	if (
		!useThisRule(
			$params->roles,
			$params->templates
		)
	){
		return;
	}
	
	global $modx;
	
	if ($modx->Event->name == 'OnDocFormPrerender'){
		//The jQuery.ddMap library including
		$output .= includeJsCss(
			$modx->config['site_url'] . 'assets/plugins/managermanager/widgets/mm_ddmap/jQuery.ddMap-2.0.min.js',
			'html',
			'jQuery.ddMap',
			'2.0'
		);
		//The main js file including
		$output .= includeJsCss(
			$modx->config['site_url'] . 'assets/plugins/managermanager/widgets/mm_ddmap/jQuery.ddMM.mm_ddMap.js',
			'html',
			'jQuery.ddMM.mm_ddMap',
			'2.1'
		);
		
		$modx->Event->output($output);
	}elseif ($modx->Event->name == 'OnDocFormRender'){
		if (is_array($params->fields)){
			$params->fields = implode(
				',',
				$params->fields
			);
		}
		
		$output = 
'
//---------- mm_ddMap :: Begin -----
$j.ddMM.getFieldElems({fields: "' . $params->fields . '"}).mm_ddMap({
	hideField: ' . intval($params->hideOriginalInput) . ',
	width: "' . $params->mapWidth . '",
	height: "' . $params->mapHeight . '"' . (
		!empty($params->defaultZoom) ?
		', defaultZoom: ' . intval($params->defaultZoom) :
		''
	) . (
		!empty($params->defaultPosition) ?
		', defaultPosition: "' . $params->defaultPosition . '"' :
		''
	) . '
});
//---------- mm_ddMap :: End -----
';
		
		$modx->Event->output($output);
	}
}
?>